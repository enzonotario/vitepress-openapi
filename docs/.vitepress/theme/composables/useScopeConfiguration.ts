import { ref } from 'vue'

const currentScope = ref('global')

export function useScopeConfiguration() {
  return {
    currentScope,
  }
}
