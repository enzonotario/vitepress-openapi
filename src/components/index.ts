import OASpec from './Common/OASpec.vue'
import OAOperation from './Operation/OAOperation.vue'
import OAHeading from './Common/OAHeading.vue'
import OAFooter from './Common/OAFooter.vue'
import OACodeBlock from './Common/OACodeBlock.vue'
import OAMarkdown from './Common/OAMarkdown.vue'
import OAPath from './Path/OAPath.vue'
import OAPaths from './Path/OAPaths.vue'
import OAPathsByTags from './Path/OAPathsByTags.vue'
import OAPathEndpoint from './Path/OAPathEndpoint.vue'
import OAResponses from './Response/OAResponses.vue'
import OAResponse from './Response/OAResponse.vue'
import OAParameter from './Parameter/OAParameter.vue'
import OAParameters from './Parameter/OAParameters.vue'
import OATryItButton from './Try/OATryItButton.vue'
import OATryWithVariables from './Try/OATryWithVariables.vue'
import OACodeSamples from './Sample/OACodeSamples.vue'
import OARequestBody from './Request/OARequestBody.vue'
import OASchemaPropertyAttributes from './Schema/OASchemaPropertyAttributes.vue'
import OASchemaTabs from './Schema/OASchemaTabs.vue'
import OASecurity from './Security/OASecurity.vue'
import OAInfo from './Common/OAInfo.vue'
import OAServers from './Common/OAServers.vue'

export {
  OASpec,
  OAOperation,
  OAHeading,
  OAFooter,
  OACodeBlock,
  OAMarkdown,
  OAPath,
  OAPaths,
  OAPathsByTags,
  OAPathEndpoint,
  OAResponses,
  OAResponse,
  OAParameter,
  OAParameters,
  OATryItButton,
  OATryWithVariables,
  OACodeSamples,
  OARequestBody,
  OASchemaPropertyAttributes,
  OASchemaTabs,
  OASecurity,
  OAInfo,
  OAServers,
}
