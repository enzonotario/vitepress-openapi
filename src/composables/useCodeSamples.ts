const availableLanguages = [
  {
    lang: 'curl',
    label: 'cURL',
    highlighter: 'bash',
  },
  {
    lang: 'javascript',
    label: 'JavaScript',
    highlighter: 'javascript',
  },
  {
    lang: 'php',
    label: 'PHP',
    highlighter: 'php',
  },
  {
    lang: 'python',
    label: 'Python',
    highlighter: 'python',
  },
]

export function useCodeSamples() {
  return {
    availableLanguages,
  }
}
