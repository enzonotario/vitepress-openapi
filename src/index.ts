import { OpenAPIV3 } from '@scalar/openapi-types'

export { usePaths } from './composables/usePaths'
export { useSidebar } from './composables/useSidebar'
export { OpenApi } from './lib/OpenApi'

export const httpVerbs: readonly OpenAPIV3.HttpMethods[] = Object.values(OpenAPIV3.HttpMethods)

export const literalTypes = ['string', 'number', 'integer', 'boolean', 'null']
