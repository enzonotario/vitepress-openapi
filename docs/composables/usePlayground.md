# `usePlayground` Composable

The `usePlayground` composable provides functions to manage the Playground.

You can use the `usePlayground` composable to configure the Playground in your `.vitepress/theme/index.js` file, or in any `.md` page/file.

```ts
import { usePlayground } from 'vitepress-openapi/client'

export default {
    async enhanceApp({ app, router, siteData }) {
        const playground = usePlayground()
        
        // Set custom security scheme default values.
        playground.setSecuritySchemeDefaultValues({
            'http-basic': 'Custom Basic Auth',
            'http-bearer': 'Custom Bearer Token',
            // ...
        })
    }
}
```

## Security Scheme Default Values Configuration

| Function                         | Description                                      | Default Value                                                                                                                                 | Allowed Values                         |
|----------------------------------|--------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------|
| `setSecuritySchemeDefaultValues` | Sets custom default values for security schemes. | `{ 'http-basic': 'Basic Auth', 'http-bearer': 'Token', 'apiKey': null, 'openIdConnect': 'OpenID Connect', 'oauth2': 'OAuth2 Token', }` | Partial\<SecuritySchemeDefaultValues\> |
