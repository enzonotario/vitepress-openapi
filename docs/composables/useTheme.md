---
prev:
  text: 'Sidebar Items'
  link: /sidebar/sidebar-items
---

# `useTheme` composable

The `useTheme` composable provides functions to configure the theme.

You can use the `useTheme` composable to configure the theme in your `.vitepress/theme/index.js` file, or in any `.md` page/file.

```ts
import { useTheme, locales } from 'vitepress-openapi/client'

export default {
    async enhanceApp({app, router, siteData}) {
        useTheme({
            requestBody: {
                // Set the default schema view.
                defaultView: 'schema', // schema or contentType
            },
            jsonViewer: {
                // Set the JSON viewer depth.
                deep: Infinity,
                // Set the JSON viewer renderer.
                renderer: 'vue-json-pretty', // vue-json-pretty or shiki
            },
            schemaViewer: {
                // Set the schema viewer depth.
                deep: Infinity,
            },
            // Set the heading levels.
            headingLevels: {
                h1: 1,
                h2: 2,
                h3: 3,
                h4: 4,
                h5: 5,
                h6: 6,
            },
            response: {
                // Set the response code selector.
                codeSelector: 'tabs', // tabs or select
                // Set the maximum number of tabs, after which a Select will be shown.
                codeMaxTabs: 5,
                body: {
                    // Set the default view.
                    defaultView: 'schema', // schema or contentType
                },
            },
            playground: {
                jsonEditor: {
                    // Set the mode of the JSON editor.
                    mode: 'tree', // text, tree, or table
                    // Set the visibility of the main menu bar.
                    mainMenuBar: false,
                    // Set the visibility of the navigation bar.
                    navigationBar: false,
                    // Set the visibility of the status bar.
                    statusBar: false,
                },
            },
            operation: {
                // Set the operation badges. The order is respected.
                badges: ['deprecated'],
                // Slots to render in the OAOperation component.
                slots: [
                    'header',
                    'path',
                    'description',
                    'security',
                    'request-body',
                    'responses',
                    'playground',
                    'code-samples',
                    'branding',
                    'footer',
                ],
                // Slots to hide in the OAOperation component.
                hiddenSlots: [],
                // Set the number of columns to use in the OAOperation component.
                cols: 2,
                // Set the default base URL.
                defaultBaseUrl: 'http://localhost',
                // Deprecated. Use `server.getServers` instead.
                getServers: ({ method, path, operation }) => Array<string>,
            },
            // Set the i18n configuration.
            i18n: {
                locale: 'en', // en | es | ja | pt-BR | string
                fallbackLocale: 'en', // en | es | ja | pt-BR | string
                messages: {
                    en: {
                        ...locales.en,
                        'operation.badgePrefix.operationId': 'Operation ID',
                    },
                    es: {
                        ...locales.es,
                        'operation.badgePrefix.operationId': 'ID de operación',
                    },
                },
                availableLocales: [
                    { code: 'en', label: 'English' },
                    { code: 'es', label: 'Español' },
                    { code: 'ja', label: 'Japanese' },
                    { code: 'pt-BR', label: 'Português (Brasil)' },
                ],
            },
            // Set spec configuration.
            spec: {
                groupByTags: true, // Group paths by tags.
                collapsePaths: false, // Collapse paths when grouping by tags.
                showPathsSummary: true, // Show a summary of the paths when grouping by tags.
                avoidCirculars: false, // Avoid circular references when parsing schemas.
                lazyRendering: false, // Lazy render Paths and Tags components.
                defaultTag: 'Default', // Default tag to use when a path has no tags.
                wrapExamples: true, // Wrap examples in a row or show them in a column.
                disableDownload: false, // Disable the download button in the info section.
            },
            server: {
                // Set a custom function to get servers.
                getServers: ({ method, path, operation }) => Array<string>,
                // Allow custom servers.
                allowCustomServer: true,
            },
            markdown: {
                operationLink: {
                    // Link prefix to search for operations.
                    linkPrefix: '/operations/',
                    // Transform the href of operation links.
                    transformHref: (href) => {
                        return `/example${href}`
                    },
                },
            },
        })
    }
}
```

## Schema Configuration

| Function               | Description                         | Default Value | Allowed Values              |
|------------------------|-------------------------------------|---------------|-----------------------------|
| `setShowBaseURL`       | Sets whether the base URL is shown. | `true`        | `true`, `false`             |

## Request Body Configuration

| Function                    | Description                   | Default Value | Allowed Values              |
|-----------------------------|-------------------------------|---------------|-----------------------------|
| `setRequestBodyDefaultView` | Sets the default schema view. | `'schema'`    | `'schema'`, `'contentType'` |

## JSON Viewer Configuration

| Function                | Description                    | Default Value       | Allowed Values                 |
|-------------------------|--------------------------------|---------------------|--------------------------------|
| `setJsonViewerDeep`     | Sets the JSON viewer depth.    | `Infinity`          | `number`                       |
| `setJsonViewerRenderer` | Sets the JSON viewer renderer. | `'vue-json-pretty'` | `'vue-json-pretty'`, `'shiki'` |

## Schema Viewer Configuration

| Function              | Description                   | Default Value | Allowed Values |
|-----------------------|-------------------------------|---------------|----------------|
| `setSchemaViewerDeep` | Sets the schema viewer depth. | `Infinity`    | `number`       |

## Heading Levels Configuration

| Function           | Description              | Default Value                                  | Allowed Values                                                               |
|--------------------|--------------------------|------------------------------------------------|------------------------------------------------------------------------------|
| `setHeadingLevels` | Sets the heading levels. | `{ h1: 1, h2: 2, h3: 3, h4: 4, h5: 5, h6: 6 }` | `{ h1: number, h2: number, h3: number, h4: number, h5: number, h6: number }` |

## Response Configuration

| Function                     | Description                                                          | Default Value | Allowed Values              |
|------------------------------|----------------------------------------------------------------------|---------------|-----------------------------|
| `setResponseCodeSelector`    | Sets the response code selector.                                     | `'tabs'`      | `'tabs'`, `'select'`        |
| `setResponseCodeMaxTabs`     | Sets the maximum number of tabs, after which a Select will be shown. | `5`           | `number`                    |
| `setResponseBodyDefaultView` | Sets the default view of the response body.                          | `'schema'`    | `'schema'`, `'contentType'` |

## Playground JSON Editor Configuration

| Function                               | Description                                | Default Value | Allowed Values                |
|----------------------------------------|--------------------------------------------|---------------|-------------------------------|
| `setPlaygroundJsonEditorMode`          | Sets the mode of the JSON editor.          | `'tree'`      | `'text'`, `'tree'`, `'table'` |
| `setPlaygroundJsonEditorMainMenuBar`   | Sets the visibility of the main menu bar.  | `false`       | `true`, `false`               |
| `setPlaygroundJsonEditorNavigationBar` | Sets the visibility of the navigation bar. | `false`       | `true`, `false`               |
| `setPlaygroundJsonEditorStatusBar`     | Sets the visibility of the status bar.     | `false`       | `true`, `false`               |

## Operation Configuration

| Function                     | Description                                        | Default Value        | Allowed Values                  |
|------------------------------|----------------------------------------------------|----------------------|---------------------------------|
| `setOperationBadges`         | Sets the operation badges. The order is respected. | `['deprecated']`     | `['deprecated', 'operationId']` |
| `setOperationDefaultBaseUrl` | Sets the default base URL.                         | `'http://localhost'` | `string`                        |

## I18n Configuration

| Function        | Description                  | Default Value                                               | Allowed Values                                                                                                                           |
|-----------------|------------------------------|-------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
| `setI18nConfig` | Sets the i18n configuration. | `{ locale: 'en', fallbackLocale: 'en', messages: locales }` | `{ locale: 'es' \| 'en' \| 'ja' \| 'pt-BR' \| string, fallbackLocale: 'es' \| 'en' \| 'ja' \| 'pt-BR' \| string, messages: Record<'es' \| 'en' \| 'ja' \| 'pt-BR', Record<string, Record<string, string>>> }` |

## Spec Configuration

| Function        | Description                  | Default Value                                                                                                                             | Allowed Values                                                                                                                                     |
|-----------------|------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
| `setSpecConfig` | Sets the spec configuration. | `{ groupByTags: true, collapsePaths: false, showPathsSummary: true, avoidCirculars: false, lazyRendering: false, defaultTag: 'Default', wrapExamples: true, disableDownload: false }` | `{ groupByTags: boolean, collapsePaths: boolean, showPathsSummary: boolean, avoidCirculars: boolean, lazyRendering: boolean, defaultTag: string, wrapExamples: boolean, disableDownload: boolean }` |

## Server Configuration

| Function                      | Description                              | Default Value                                                                         | Allowed Values                                                                           |
|-------------------------------|------------------------------------------|---------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------|
| `setServerConfig`             | Sets the server configuration.           | `{ getServers: ({ method, path, operation }) => Array<string>, allowCustomServer: true }` | `{ getServers: ({ method, path, operation }) => Array<string>, allowCustomServer: boolean }` |
| `getServers`                  | Gets the servers for an operation.       | `({ method, path, operation }) => Array<string>`                                           | `({ method, path, operation }) => Array<string>`                                              |
| `getServerAllowCustomServer` | Gets whether custom servers are allowed. | `true`                                                                                | `true`, `false`                                                                          |

## Markdown Configuration

| Function              | Description                       | Default Value                                                | Allowed Values                                                                                                                                                                    |
|------------------------|-----------------------------------|------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------| 
| `setMarkdownConfig`    | Sets the markdown configuration.  | `{ operationLink: { linkPrefix: '/operations/' } }`        | `{ operationLink: { linkPrefix: string, transformHref: (href: string) => string, createOperationLinkHtml: (href: string, method: string, title: string) => string } }`            |
| `getMarkdownConfig`    | Gets the markdown configuration.  | `{ operationLink: { linkPrefix: '/operations/' } }`        | `{ operationLink: { linkPrefix: string, transformHref: (href: string) => string, createOperationLinkHtml: (href: string, method: string, title: string) => string } }` |
| `getOperationLinkConfig` | Gets the operation link configuration. | `{ linkPrefix: '/operations/' }`                    | `{ linkPrefix: string, transformHref: (href: string) => string, createOperationLinkHtml: (href: string, method: string, title: string) => string }`                    |
