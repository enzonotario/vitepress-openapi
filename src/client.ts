import type { Awaitable } from 'vitepress'
import type { EnhanceAppContext, Theme } from 'vitepress/client'
import type { Component } from 'vue'
import type { OpenApi } from './lib/OpenApi'
import { createI18n } from '@byjohann/vue-i18n'
import { watch } from 'vue'
import * as components from './components'
import { DEFAULT_OPERATION_SLOTS, useTheme } from './composables/useTheme'

interface OAEnhanceAppContext extends EnhanceAppContext {
  openapi?: ReturnType<typeof OpenApi> | null
}

interface VPTheme {
  Layout: Component
  enhanceApp: (ctx: OAEnhanceAppContext) => Awaitable<void>
  extends?: Theme
}

export const theme = {
  enhanceApp({ app }) {
    const themeConfig = useTheme()

    const i18nConfig = themeConfig.getI18nConfig()

    const i18n = createI18n({
      defaultLocale: i18nConfig.locale?.value || 'en',
      messages: i18nConfig.messages || {},
    })

    app.use(i18n)

    watch(
      () => themeConfig.getLocale(),
      () => {
        i18n.setLocale(themeConfig.getLocale())
      },
      {
        immediate: true,
      },
    )

    Object.entries(components).forEach(([key, component]) => {
      app.component(key, component)
    })
  },
} as VPTheme

export { useOpenapi } from './composables/useOpenapi'
export { usePlayground } from './composables/usePlayground'
export { DEFAULT_OPERATION_SLOTS }
export { useShiki } from './composables/useShiki'
export { useTheme } from './composables/useTheme'
export { generateCodeSample } from './lib/codeSamples/generateCodeSample'
export { OARequest } from './lib/codeSamples/request'
export { OpenApi } from './lib/OpenApi'
export { locales } from './locales'
