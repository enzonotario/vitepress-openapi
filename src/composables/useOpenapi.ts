import type { OpenAPI } from '@scalar/openapi-types'
import { createOpenApiInstance } from '../lib/createOpenApiInstance'
import type { UseThemeConfigUnref } from './useTheme'
import { useTheme } from './useTheme'

export interface OpenAPIData {
  id: string
  spec: OpenAPI.Document
  openapi: any
  config: any
}

export type Schemas = Map<string, OpenAPIData>

export const DEFAULT_SCHEMA = 'main'

const schemas: Schemas = new Map()

let mainSchema: OpenAPI.Document | null = null

export function useOpenapi({
  spec,
  config,
}: {
  spec?: OpenAPI.Document
  config?: UseThemeConfigUnref
} = {}) {
  if (config) {
    useTheme(config)
  }

  if (spec !== null && spec !== undefined) {
    setupOpenApi({ spec, config })
  }

  /**
   * @deprecated Use `useOpenapi({ spec })` instead.
   */
  function setSpec(value: OpenAPI.Document) {
    console.warn('Deprecated usage of `setSpec`. Use `useOpenapi({ spec })` instead.')
    setupOpenApi({ spec: value })
  }

  function setupOpenApi({ spec, config }: { spec: OpenAPI.Document, config?: UseThemeConfigUnref }) {
    addSchema({ id: DEFAULT_SCHEMA, spec, config })
    mainSchema = schemas.get(DEFAULT_SCHEMA) as OpenAPI.Document
  }

  function addSchema({ id, spec, config }: { id: string, spec: OpenAPI.Document, config?: UseThemeConfigUnref | null }) {
    const openapi = createOpenApiInstance({ spec })

    // @ts-expect-error: This adds all the properties of the OpenAPI instance to the schema.
    schemas.set(id, {
      ...openapi,
      id,
      config,
    })
  }

  return {
    ...mainSchema,
    json: mainSchema?.spec,
    setSpec,
    schemas,
  }
}
