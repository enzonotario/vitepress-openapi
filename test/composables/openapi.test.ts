import { describe, expect, it } from 'vitest'
import { OpenApi } from '../../src'
import { useOpenapi } from '../../src/composables/useOpenapi'
import { spec, specWithSchemaAndContentTypes } from '../testsConstants'

describe('openapi with spec', () => {
  const openapi = OpenApi({ spec })

  it('setSpec and getSpec work correctly', () => {
    const customOpenapi = OpenApi()

    expect(customOpenapi.getSpec()).toEqual({})

    customOpenapi.setSpec(spec)
    expect(customOpenapi.getSpec()).toEqual(spec)

    const newSpec = { openapi: '3.0.0', info: { title: 'Custom API', version: '2.0.0' } }
    customOpenapi.setSpec(newSpec)
    expect(customOpenapi.getSpec()).toEqual(newSpec)
  })

  it('stores and retrieves originalSpec correctly', () => {
    const originalSpec = { openapi: '3.0.0', info: { title: 'Original API', version: '1.0.0' } }
    const parsedSpec = { openapi: '3.0.0', info: { title: 'Parsed API', version: '1.0.0' } }

    const customOpenapi = OpenApi({
      spec: parsedSpec,
      originalSpec,
    })

    expect(customOpenapi.getSpec()).toEqual(parsedSpec)
    expect(customOpenapi.getOriginalSpec()).toEqual(originalSpec)

    const newOriginalSpec = { openapi: '3.0.0', info: { title: 'New Original API', version: '2.0.0' } }
    customOpenapi.setOriginalSpec(newOriginalSpec)
    expect(customOpenapi.getOriginalSpec()).toEqual(newOriginalSpec)
  })

  it('returns the correct operation for getOperation', () => {
    const result = openapi.getOperation('getUsers')
    expect(result).toEqual(spec.paths['/users'].get)
  })

  it('returns the correct path for getOperationPath', () => {
    const result = openapi.getOperationPath('getUsers')
    expect(result).toBe('/users')
  })

  it('returns the correct parameters for getOperationParameters', () => {
    const result = openapi.getOperationParameters('getUsers')
    expect(result).toEqual(spec.paths['/users'].get.parameters)
  })

  it('returns the correct tags for getTags', () => {
    const result = openapi.getOperationsTags()
    expect(result).toEqual(['users', 'pets'])
  })

  it('returns the correct info for getInfo', () => {
    const result = openapi.getInfo()
    expect(result).toEqual(spec.info)
  })

  it('returns the correct external docs for getExternalDocs', () => {
    const result = openapi.getExternalDocs()
    expect(result).toEqual(spec.externalDocs)
  })

  it('returns the correct servers for getServers', () => {
    const result = openapi.getServers()
    expect(result).toEqual(spec.servers)
  })

  it('getOperationsTags returns all unique tags', () => {
    const api = useOpenapi({
      spec: {
        ...spec,
        paths: {
          '/users': {
            get: {
              operationId: 'getUsers',
              tags: ['users'],
            },
          },
          '/profile': {
            get: {
              operationId: 'getProfile',
              tags: ['profile'],
            },
          },
        },
      },
    })
    const tags = api.getOperationsTags()
    expect(tags).toEqual(['users', 'profile'])
  })

  it('getOperationsTags returns empty array if no paths', () => {
    const api = useOpenapi({ spec: { openapi: '3.0.0' } })
    const tags = api.getOperationsTags()
    expect(tags).toEqual([])
  })

  it('getPathsByTags returns paths with specified tags', () => {
    const api = useOpenapi({
      spec: {
        ...spec,
        paths: {
          '/users': {
            get: {
              operationId: 'getUsers',
              tags: ['users'],
            },
          },
          '/profile': {
            get: {
              operationId: 'getProfile',
              tags: ['profile'],
            },
          },
        },
      },
    })
    const paths = api.getPathsByTags('users')
    expect(paths).toMatchObject({
      '/users': {
        get: {
          operationId: 'getUsers',
          summary: 'GET /users',
          tags: ['users'],
        },
      },
    })
  })

  it('getPathsByTags returns empty object if no matching tags', () => {
    const api = useOpenapi({ spec })
    const paths = api.getPathsByTags('nonexistent')
    expect(paths).toMatchObject({})
  })

  it('sets defaultTag for operations without tags', () => {
    const api = useOpenapi({
      spec: {
        ...spec,
        paths: {
          '/users': {
            get: {
              operationId: 'getUsers',
              tags: ['users'],
            },
          },
          '/no-tags': {
            get: {
              operationId: 'getNoTags',
            },
          },
        },
      },
    })

    expect(api.getPathsWithoutTags()).toMatchObject({})

    expect(api.getOperationsTags()).toEqual(['users', 'Default'])
  })

  it('getPathsWithoutTags returns empty array if no paths without tags', () => {
    const api = useOpenapi({ spec: { openapi: '3.0.0', paths: {} } })
    const paths = api.getPathsWithoutTags()
    expect(paths).toMatchObject({})
  })

  it('getTags returns tags object', () => {
    const api = useOpenapi({ spec })
    const tags = api.getTags()
    expect(tags).toEqual([
      {
        name: 'users',
        description: 'Operations about users',
      },
      {
        name: 'pets',
        description: 'Operations about pets',
      },
      {
        name: 'Default',
        description: '',
      },
    ])
  })

  it('getTags returns empty array if no tags in spec', () => {
    const api = useOpenapi({ spec: { openapi: '3.0.0', paths: {} } })
    const tags = api.getTags()
    expect(tags).toEqual([])
  })

  it('getFilteredTags returns tags filtered by operations', () => {
    const filteredTags = openapi.getFilteredTags()

    expect(filteredTags).toEqual([
      {
        name: 'users',
        description: 'Operations about users',
      },
      {
        name: 'pets',
        description: 'Operations about pets',
      },
    ])
  })

  it('getFilteredTags handles tags not defined in spec.tags', () => {
    const customSpec = {
      openapi: '3.0.0',
      paths: {
        '/test': {
          get: {
            operationId: 'test',
            tags: ['customTag'],
          },
        },
      },
    }
    const api = useOpenapi({ spec: customSpec })
    const filteredTags = api.getFilteredTags()

    expect(filteredTags).toEqual([
      {
        name: 'customTag',
        description: null,
      },
    ])
  })

  it('returns the correct operation servers for getOperationServers', () => {
    const result = openapi.getOperationServers('getUsers')
    expect(result).toEqual([
      ...spec.paths['/users'].get.servers,
      ...spec.servers,
    ])
  })

  it('getOperationByMethodAndPath returns the correct operation', () => {
    const result = openapi.getOperationByMethodAndPath('get', '/users')
    expect(result).toEqual(spec.paths['/users'].get)
  })

  it('getOperationByMethodAndPath returns null for non-existing operation', () => {
    const result = openapi.getOperationByMethodAndPath('get', '/non-existing')
    expect(result).toBeNull()
  })
})

describe('spec with different servers for specific path', () => {
  const spec = {
    openapi: '3.0.0',
    servers: [
      {
        url: 'https://api.example.com',
      },
    ],
    paths: {
      '/use-global-server': {
        get: {
          operationId: 'useGlobalServer',
        },
      },
      '/use-path-server': {
        get: {
          operationId: 'usePathServer',
        },
        servers: [
          {
            url: 'https://api.path.com',
          },
        ],
      },
      '/use-local-server': {
        get: {
          operationId: 'useLocalServer',
          servers: [
            {
              url: 'https://api.local.com',
            },
          ],
        },
      },
    },
  }

  const openapi = useOpenapi({ spec })

  it('returns correct servers for getOperationServers', () => {
    const useGlobalServer = openapi.getOperationServers('useGlobalServer')
    expect(useGlobalServer).toEqual([
      {
        url: 'https://api.example.com',
      },
    ])

    const usePathServer = openapi.getOperationServers('usePathServer')
    expect(usePathServer).toEqual([
      {
        url: 'https://api.path.com',
      },
      {
        url: 'https://api.example.com',
      },
    ])

    const useLocalServer = openapi.getOperationServers('useLocalServer')
    expect(useLocalServer).toEqual([
      {
        url: 'https://api.local.com',
      },
      {
        url: 'https://api.example.com',
      },
    ])
  })
})

describe('custom specs', () => {
  it('supports custom default tag name and description', () => {
    const customTag = 'Custom'
    const customDescription = 'Custom default description'
    const openapi = useOpenapi({
      spec: {
        openapi: '3.0.0',
        paths: {
          '/no-tags': {
            get: {
              operationId: 'getNoTags',
            },
          },
        },
      },
      config: {
        spec: {
          defaultTag: customTag,
          defaultTagDescription: customDescription,
        },
      },
    })

    expect(openapi.getOperationsTags()).toEqual([customTag])

    const tags = openapi.getTags()

    expect(tags).toEqual([
      {
        name: customTag,
        description: customDescription,
      },
    ])
  })
})

describe('schemaParser', async () => {
  const openapi = await useOpenapi().async({
    spec: specWithSchemaAndContentTypes,
  })

  it('parses schema with content types', () => {
    const getPetsOperation = openapi.getOperation('getPets')

    const schemaUi = getPetsOperation.responses['400'].content['application/json'].ui

    expect(schemaUi).toMatchSnapshot()
  })
})

describe('operationParsed -> securityUi', async () => {
  const openapi = await useOpenapi().async({
    spec: {
      openapi: '3.0.0',
      paths: {
        '/onlyApiKey': {
          get: {
            operationId: 'onlyApiKey',
            security: [
              {
                apiKey: [],
              },
            ],
          },
        },
        '/onlyBearerAuth': {
          get: {
            operationId: 'onlyBearerAuth',
            security: [
              {
                bearerAuth: [],
              },
            ],
          },
        },
        '/apiKeyAndBearerAuth': {
          get: {
            operationId: 'apiKeyAndBearerAuth',
            security: [
              {
                apiKey: [],
                bearerAuth: [],
              },
            ],
          },
        },
        '/apiKeyOrBearerAuth': {
          get: {
            operationId: 'apiKeyOrBearerAuth',
            security: [
              {
                apiKey: [],
              },
              {
                bearerAuth: [],
              },
            ],
          },
        },
        '/noSecurity': {
          get: {
            operationId: 'noSecurity',
            security: [],
          },
        },
      },
      components: {
        securitySchemes: {
          apiKey: {
            type: 'apiKey',
            name: 'api_key',
            in: 'header',
          },
          bearerAuth: {
            type: 'http',
            scheme: 'bearer',
          },
        },
      },
    },
  })

  it('returns correct securityUi for operations', () => {
    const onlyApiKey = openapi.getOperation('onlyApiKey').securityUi
    expect(onlyApiKey).toEqual([
      {
        id: 'apiKey',
        schemes: {
          apiKey: {
            type: 'apiKey',
            name: 'api_key',
            in: 'header',
          },
        },
      },
    ])

    const onlyBearerAuth = openapi.getOperation('onlyBearerAuth').securityUi
    expect(onlyBearerAuth).toEqual([
      {
        id: 'bearerAuth',
        schemes: {
          bearerAuth: {
            type: 'http',
            scheme: 'bearer',
          },
        },
      },
    ])

    const apiKeyAndBearerAuth = openapi.getOperation('apiKeyAndBearerAuth').securityUi
    expect(apiKeyAndBearerAuth).toEqual([
      {
        id: 'apiKey|bearerAuth',
        schemes: {
          apiKey: {
            type: 'apiKey',
            name: 'api_key',
            in: 'header',
          },
          bearerAuth: {
            type: 'http',
            scheme: 'bearer',
          },
        },
      },
    ])

    const apiKeyOrBearerAuth = openapi.getOperation('apiKeyOrBearerAuth').securityUi
    expect(apiKeyOrBearerAuth).toEqual([
      {
        id: 'apiKey',
        schemes: {
          apiKey: {
            type: 'apiKey',
            name: 'api_key',
            in: 'header',
          },
        },
      },
      {
        id: 'bearerAuth',
        schemes: {
          bearerAuth: {
            type: 'http',
            scheme: 'bearer',
          },
        },
      },
    ])
  })
})
