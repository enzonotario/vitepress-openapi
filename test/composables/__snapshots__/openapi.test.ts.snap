// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schemaParser > parses schema with content types 1`] = `
{
  "description": "RFC 7807 (https://datatracker.ietf.org/doc/html/rfc7807)",
  "name": "",
  "properties": [
    {
      "examples": [
        "https://example.com/errors/generic-error",
      ],
      "name": "type",
      "required": false,
      "types": [
        "string",
      ],
    },
    {
      "examples": [
        "Something went wrong here.",
      ],
      "name": "title",
      "required": false,
      "types": [
        "string",
      ],
    },
    {
      "constraints": {
        "format": "int64",
      },
      "examples": [
        403,
      ],
      "name": "status",
      "required": false,
      "types": [
        "integer",
      ],
    },
    {
      "examples": [
        "Unfortunately, we canâ€™t provide further information.",
      ],
      "name": "detail",
      "required": false,
      "types": [
        "string",
      ],
    },
  ],
  "required": false,
  "types": [
    "object",
  ],
}
`;
